assets/js/atomic/utils/render-parent-block.tsx

```diff
@@ -43,53 +111,46 @@ const renderInnerBlocks = ( {
 	depth = 1,
 	children,
 }: renderInnerBlockProps ): ( JSX.Element | null )[] | null => {
-	return Array.from( children ).map( ( el: Element, index: number ) => {
+	if ( ! children || children.length === 0 ) {
+		return null;
+	}
+	return Array.from( children ).map( ( element: Element, index: number ) => {
 		const { blockName = '', ...componentProps } = {
 			key: `${ parentBlockName }_${ depth }_${ index }`,
-			...( el instanceof HTMLElement ? el.dataset : {} ),
+			...( element instanceof HTMLElement ? element.dataset : {} ),
 		};
 
-		const componentChildren =
-			el.children && el.children.length
-				? renderInnerBlocks( {
-						children: el.children,
-						blockName: parentBlockName,
-						blockMap,
-						depth: depth + 1,
-						blockWrapper,
-				  } )
-				: null;
-
-		const LayoutComponent =
-			blockName && blockMap[ blockName ]
-				? ( blockMap[ blockName ] as React.ElementType )
-				: null;
-
-		if ( ! LayoutComponent ) {
-			const element = parse( el.outerHTML );
-
-			if ( isValidElement( element ) ) {
-				return componentChildren
-					? cloneElement( element, componentProps, componentChildren )
-					: cloneElement( element, componentProps );
-			}
-			return null;
-		}
+		const InnerBlockComponent = getInnerBlockComponent(
+			blockName,
+			blockMap,
+			element
+		);
 
-		const LayoutComponentWrapper = ( blockWrapper
+		const InnerBlockComponentWrapper = blockWrapper
 			? blockWrapper
-			: Fragment ) as React.ElementType;
+			: Fragment;
 
 		return (
 			<Suspense
 				key={ `${ parentBlockName }_${ depth }_${ index }_suspense` }
 				fallback={ <div className="wc-block-placeholder" /> }
 			>
-				<LayoutComponentWrapper>
-					<LayoutComponent { ...componentProps }>
-						{ componentChildren }
-					</LayoutComponent>
-				</LayoutComponentWrapper>
+				<InnerBlockComponentWrapper>
+					<InnerBlockComponent { ...componentProps }>
+						{ renderInnerBlocks( {
+							children: element.children,
+							blockName: parentBlockName,
+							blockMap,
+							depth: depth + 1,
+							blockWrapper,
+						} ) }
+						{ renderForcedBlocks(
+							blockName,
+							blockMap,
+							element.children
+						) }
+					</InnerBlockComponent>
+				</InnerBlockComponentWrapper>
```

Comment:
This got even more complicated you’re now the only one who can maintain it so we need to document what's happening here later.

Reply:
Added more explanation in https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4671/commits/47b4f328807cc05dae426130e21bbfd2489d733c to make it easier to follow.